source/YTEPs/YTEP-0038.rst

```diff
@@ -0,0 +1,87 @@
+YTEP-0038: Type Annotations
+===========================
+
+Abstract
+--------
+
+Created: March 1, 2021
+Author: Matthew Turk
+
+This YTEP proposes that we undertake a staged process of adding [type
+annotations](https://docs.python.org/3/library/typing.html) to the yt
+codebase.
+
+Status
+------
+
+Proposed, after havingbeen discussed.
+
+Project Management Links
+------------------------
+
+  * [PR 3064 - Add a `_typing` module](https://github.com/yt-project/yt/pull/3064)
+  * [PR 3028 - pyannotate for type hints](https://github.com/yt-project/yt/pull/3028)
+  * [PR 2917 - Drop support for Py 3.6](https://github.com/yt-project/yt/pull/2917)
+  * [PR 2352 - Experiment with type hinting](https://github.com/yt-project/yt/pull/2352)
+  * [Somewhat related YTEP on traitlets](https://github.com/yt-project/ytep/pull/8)
+
+Detailed Description
+--------------------
+
+In the 3.x series, Python had added support for type annotations/hinting.
+These are optional annotations that specify what types a given function
+expects to receive or to emit.
+
+This allows a number of useful features, chief among them a more rigorous
+method for error-checking in code and a way to provide much better
+autocompletion and IDE functionality for developing both scripts *and* code
+within yt.  If your IDE is aware of what kinds of objects will be returned by 
+a function, it can then provide options about what you might be able to do
+with that object. A recent summary (as of this writing) can be found
+[here](https://www.augmentedmind.de/2020/10/11/static-python-type-hints/).
+
+There are also possibilities of future optimizations that can build on type
+hinting, for instance in tools like Cython and Numba, but that is not
+considered a top-tier reason for adding type hints.
+
+However, there are some features of type-hinting that would be extremely
+useful to yt that only started appearing in Python 3.7. Part of this
```

Comment:
As Corentin said, it actually started with Python 3.6
the builtin `typing` and `types` (as well as `collections` in Python 3.9) modules have be getting more and more support each version, and it's true that dropping Python 3.6 makes the task a little less difficult, but it's not a requirement to get started.

For completeness, the specific improvement in Python 3.7 over 3.6 is that type hints evaluation is postponed, instead of immediate. This makes the following snippet invalid in Python 3.6 but valid in 3.7 and above

```python
def baconify(spam) -> Bacon:  # error, Bacon type isn't defined yet !
    return Bacon(spam)

class Bacon:
    def __init__(self, spam):
        ...
```

Note that is order to make this Python 3.6 compliant, it is sufficient to quote the problematic annotation as

```python
def baconify(spam) -> "Bacon":
```
The main downside is that it creates unnecessary churn to write Python 3.6 compliant annotation if we're going to drop it immediately after, which is why those two problems should indeed be discussed together.